src/callbacks_step/amr_dg2d.jl

```diff
@@ -10,9 +10,8 @@ function refine!(u_ode::AbstractVector, adaptor, mesh::TreeMesh{2},
   # Determine for each existing element whether it needs to be refined
   needs_refinement = falses(nelements(dg, cache))
 
-  # The "Ref(...)" is such that we can vectorize the search but not the array that is searched
-  elements_to_refine = searchsortedfirst.(Ref(cache.elements.cell_ids[1:nelements(dg, cache)]),
-                                          cells_to_refine)
+  # Find all indices of elements whose cell ids are in cells_to_refine
+  elements_to_refine = findall(cell_id -> cell_id in cells_to_refine, cache.elements.cell_ids)
```

Comment:
Firstly, `cell_ids` is a field of the element container, so for an element id it gives the corresponding cell id. Its length should always be `nelements(dg, cache)`.

Secondly, the `searchsortedfirst` version always outputs an array of the same size as `cells_to_refine`. When running serially, this is fine. However, in parallel execution `cells_to_refine` can be bigger.

Let me give you an example:
I have 16 elements, 8 of them are on rank 0, 4 on rank 1, 4 on rank 2. `cells_to_refine` is `[3,  4,  5,  6,  8,  9,  10,  11,  13,  14,  15,  16,  18,  19,  20,  21]` on all ranks.
Rank 0 has the `cell_ids = [3,  4,  5,  6,  8,  9,  10,  11]`, so all elements of this rank should be refined. I expect `elements_to_refine = 1:8`, which is what `findall` returns. However, `searchsortedfirst` returns `[1,  2,  3,  4,  5,  6,  7,  8,  9,  9,  9,  9,  9,  9,  9,  9]`.
The same happens on rank 2: `cell_ids = [13,  14,  15,  16]`, so the correct result would be `1:4`. `searchsortedfirst` returns `[1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  3,  4,  5,  5,  5,  5]`.

I guess, one could modify the `searchsortedfirst` version to get the expected output (probably with a better performance than `findall`), but I really like the readability of the short `findall` version. Perhaps, we can find a more readable version which uses the sorting of the array to search.

Reply:
OK, good enough for me. However, maybe we should create an issue to investigate the performance difference of `findall` vs `searchsortedfirst`? I think it doesn't make a difference for the toy problems we have here, but maybe with larger simulations it will be noticable. That way, we won't forget (e.g., in case someone notices that the AMR algorithm is sluggish).

Reply:
I will create a new issue once this PR is merged (in case we find a better version first).